### ISSUE TITLE
Concern about the speed

### ISSUE THREAD

**Issue by haofanwang** [AndyShih12/paradigms#4]
It seems that paradigms is only support for DDPM and DDIM. When testing on 2*V100 with batch_size=10 and DDPM, it achieves 1.5 times speed up. But if I use UniPCMultistepScheduler, the time cost is similar to DDPMParallelScheduler. 

Any plan for support UniPCMultistepScheduler? @AndyShih12

**Comment by AndyShih12** [comment]
Thanks for trying out our code.

To get the most increase in speed, it is important that there is enough parallel computing power. This means:
1. Using beefier GPUs (e.g. A100)
2. Using more GPUs (using 8 will be faster than 2)
3. Having faster interconnect between the GPUs (e.g. A100 SXM will be faster than A100 PCIe)

Based on our experience, that should get you from 1.5x speedup to 3x speedup.

-----

In terms of other schedulers, currently Diffusers only has ParaDDPM and ParaDDIM. In this repo, we additionally have ParaDPMSolver. For the moment, you can use it like this:

```python
import torch
from diffusers import DDPMParallelScheduler
from diffusers import StableDiffusionParadigmsPipeline

from paradigms.paradpmsolver_scheduler import ParaDPMSolverMultistepScheduler

scheduler = ParaDPMSolverMultistepScheduler.from_pretrained("runwayml/stable-diffusion-v1-5", subfolder="scheduler", timestep_spacing="trailing")
scheduler._is_ode_scheduler = True
scheduler._is_parallel = True
scheduler.config.thresholding = False

pipe = StableDiffusionParadigmsPipeline.from_pretrained(
    "runwayml/stable-diffusion-v1-5", scheduler=scheduler, torch_dtype=torch.float16
)
pipe = pipe.to("cuda")
ngpu, batch_per_device = torch.cuda.device_count(), 5
pipe.wrapped_unet = torch.nn.DataParallel(pipe.unet, device_ids=[d for d in range(ngpu)])
prompt = "a photo of an astronaut riding a horse on mars"

image = pipe(prompt, parallel=ngpu * batch_per_device, num_inference_steps=200).images[0]
image.save("image.png")
```

-----

To support other schedulers like UniPCMultistepScheduler, it should be quite easy. Just implement a tensorized version of step() function like [here](https://github.com/AndyShih12/paradigms/blob/47b870c766408fd8d1c3e8c78c12a369d462200f/paradigms/paradpmsolver_scheduler.py#L23C17-L23C17) and a get_varance() function like [here](https://github.com/AndyShih12/paradigms/blob/47b870c766408fd8d1c3e8c78c12a369d462200f/paradigms/paradpmsolver_scheduler.py#L11C12-L11C12). The three examples of ParaDDPM, ParaDDIM, and ParaDPMScheduler in https://github.com/AndyShih12/paradigms/tree/master/paradigms should provide enough guidance.

**Comment by haofanwang** [comment]
I found that when use DDIM or UniPCMultistepScheduler with steps=20, the speed of ParaDDIM is even slower than original scheduler. Is it normal? @AndyShih12

**Comment by AndyShih12** [comment]
Yes, in the best case, 25 steps is around the limit of when you can still see speedups. You won't see speedups if you use fewer steps than 25.

We have a recent improvement that gives good speedups for 50 steps, so stay tuned.
